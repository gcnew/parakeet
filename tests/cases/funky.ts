
import { test } from '../test_util'

import {
    char, integer, ws, asciiId, stringChoice,

    StringMismatch,

    StringParser
} from '../../src/string_combinators'

import {
    map, mapError, alt3, many, maybe, pconst, combine, combine3, combine4, terminated,

    oneOrMore as many1
} from '../../src/parser_combinators'


/* Types */
type ArithOp = 'add' | 'sub' | 'mul' | 'div'
// type Value = { kind: 'int_value', value: number }
//            | { kind: 'function', name: string, params: string[], body: Expr }

type Expr = { kind: 'constant', value: number }
          | { kind: 'arith', op: ArithOp, left: Expr, right: Expr }
          | { kind: 'var', name: string }
          | { kind: 'application', func: Expr, args: Expr[] }

type Declaration = { kind: 'function_decl', name: string, params: string[], body: Expr }
type Program     = { kind: 'program', body: Declaration[] }


/* Constructors */
function mkConstant(value: number): Expr {
    return { kind: 'constant', value };
}

function mkArith(op: ArithOp, left: Expr, right: Expr): Expr {
    return { kind: 'arith', op, left, right };
}

function mkVar(name: string): Expr {
    return { kind: 'var', name };
}

function mkApplication(func: Expr, args: Expr[]): Expr {
    return { kind: 'application', func, args };
}

function mkFunctionDecl(name: string, params: string[], body: Expr): Declaration {
    return { kind: 'function_decl', name, params, body };
}

function mkProgram(body: Declaration[]): Program {
    return { kind: 'program', body };
}


/* Parsers */
const skipWs     = maybe(ws);
const parseWs    = many1(ws);

const parseConstant = map(integer, x => mkConstant(+x));
const parseVar      = map(asciiId, x => mkVar(x));

const parseArithOp = stringChoice({
    "+": pconst<'add'>('add'),
    "-": pconst<'sub'>('sub'),
    "*": pconst<'mul'>('mul'),
    "/": pconst<'div'>('div')
});

// FunctionDecl = Id (' ' Id)* '=' Expr ';'
const parseFunctionDecl = combine4(
    asciiId,                                                                              // name
    many(combine(parseWs, asciiId, (_, id) => id)),                                       // params
    combine4(skipWs, char('='), skipWs, st => parseExpr(st), (_1, _2, _3, body) => body), // body

    mapError(combine(skipWs, char(';'), x => x), _ => ({ kind: 'pc_error' as 'pc_error', message: 'Did you forget `;` ?' })),

    mkFunctionDecl
);

// Declartion = FunctionDecl
const parseDeclaration = parseFunctionDecl;

const parseAtom = alt3(
    parseConstant,
    parseVar,
    combine3(
        combine(char('('), skipWs, _ => _),
        st => parseExpr(st),
        combine(skipWs, char(')'), _ => _),
        (_, expr) => expr
    )
);

// Application = Atom (' ' Atom)*
const parseApplication = combine(
    parseAtom,
    maybe(many1(combine(parseWs, parseAtom, (_, a) => a))),

    (func, args) => args ? mkApplication(func, args)
                         : func
);

// Arith = Application ('+'|'_'|'*'|'/') Arith
const parseArith: StringParser<StringMismatch, Expr> = combine3(
    parseApplication,
    skipWs,
    maybe(combine3(
        parseArithOp,
        skipWs,
        st => parseArith(st),
        (op, _, right) => ({ op, right })
    )),

    (left, _, r) => r ? mkArith(r.op, left, r.right)
                      : left
);

// Expr = Arith
const parseExpr = parseArith;

// Program = Declaration+
const parseProgram = terminated(combine(
    skipWs,
    many1(combine(parseDeclaration, skipWs, x => x)),
    (_, decls) => mkProgram(decls)
));


/* Test cases */
test(parseProgram, "main = 3 + 4");
test(parseProgram, "main = 3 + 4;");
test(parseProgram, "f x y = x * y; main = f 4 5 + 11;");
test(parseProgram, "c = 9; main = c;");
test(parseProgram, "c = 42; f x = x - c; main = f 29 + c;");
test(parseProgram, "x = 20; f x = x - 9; main = f 10;");
