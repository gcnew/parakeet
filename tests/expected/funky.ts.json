{
    "kind": "left",
    "value": {
        "kind": "pc_error",
        "message": "Did you forget `;` ?"
    }
}

{
    "kind": "right",
    "value": [
        {
            "kind": "program",
            "body": [
                {
                    "kind": "function_decl",
                    "name": "main",
                    "params": [],
                    "body": {
                        "kind": "arith",
                        "op": "add",
                        "left": {
                            "kind": "constant",
                            "value": 3
                        },
                        "right": {
                            "kind": "constant",
                            "value": 4
                        }
                    }
                }
            ]
        },
        {
            "source": "main = 3 + 4;",
            "offset": 13,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "main = 3 + 4;"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        {
            "kind": "program",
            "body": [
                {
                    "kind": "function_decl",
                    "name": "f",
                    "params": [
                        "x",
                        "y"
                    ],
                    "body": {
                        "kind": "arith",
                        "op": "mul",
                        "left": {
                            "kind": "var",
                            "name": "x"
                        },
                        "right": {
                            "kind": "var",
                            "name": "y"
                        }
                    }
                },
                {
                    "kind": "function_decl",
                    "name": "main",
                    "params": [],
                    "body": {
                        "kind": "arith",
                        "op": "add",
                        "left": {
                            "kind": "application",
                            "func": {
                                "kind": "var",
                                "name": "f"
                            },
                            "args": [
                                {
                                    "kind": "constant",
                                    "value": 4
                                },
                                {
                                    "kind": "constant",
                                    "value": 5
                                }
                            ]
                        },
                        "right": {
                            "kind": "constant",
                            "value": 11
                        }
                    }
                }
            ]
        },
        {
            "source": "f x y = x * y; main = f 4 5 + 11;",
            "offset": 33,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "f x y = x * y; main = f 4 5 + 11;"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        {
            "kind": "program",
            "body": [
                {
                    "kind": "function_decl",
                    "name": "c",
                    "params": [],
                    "body": {
                        "kind": "constant",
                        "value": 9
                    }
                },
                {
                    "kind": "function_decl",
                    "name": "main",
                    "params": [],
                    "body": {
                        "kind": "var",
                        "name": "c"
                    }
                }
            ]
        },
        {
            "source": "c = 9; main = c;",
            "offset": 16,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "c = 9; main = c;"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        {
            "kind": "program",
            "body": [
                {
                    "kind": "function_decl",
                    "name": "c",
                    "params": [],
                    "body": {
                        "kind": "constant",
                        "value": 42
                    }
                },
                {
                    "kind": "function_decl",
                    "name": "f",
                    "params": [
                        "x"
                    ],
                    "body": {
                        "kind": "arith",
                        "op": "sub",
                        "left": {
                            "kind": "var",
                            "name": "x"
                        },
                        "right": {
                            "kind": "var",
                            "name": "c"
                        }
                    }
                },
                {
                    "kind": "function_decl",
                    "name": "main",
                    "params": [],
                    "body": {
                        "kind": "arith",
                        "op": "add",
                        "left": {
                            "kind": "application",
                            "func": {
                                "kind": "var",
                                "name": "f"
                            },
                            "args": [
                                {
                                    "kind": "constant",
                                    "value": 29
                                }
                            ]
                        },
                        "right": {
                            "kind": "var",
                            "name": "c"
                        }
                    }
                }
            ]
        },
        {
            "source": "c = 42; f x = x - c; main = f 29 + c;",
            "offset": 37,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "c = 42; f x = x - c; main = f 29 + c;"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        {
            "kind": "program",
            "body": [
                {
                    "kind": "function_decl",
                    "name": "x",
                    "params": [],
                    "body": {
                        "kind": "constant",
                        "value": 20
                    }
                },
                {
                    "kind": "function_decl",
                    "name": "f",
                    "params": [
                        "x"
                    ],
                    "body": {
                        "kind": "arith",
                        "op": "sub",
                        "left": {
                            "kind": "var",
                            "name": "x"
                        },
                        "right": {
                            "kind": "constant",
                            "value": 9
                        }
                    }
                },
                {
                    "kind": "function_decl",
                    "name": "main",
                    "params": [],
                    "body": {
                        "kind": "application",
                        "func": {
                            "kind": "var",
                            "name": "f"
                        },
                        "args": [
                            {
                                "kind": "constant",
                                "value": 10
                            }
                        ]
                    }
                }
            ]
        },
        {
            "source": "x = 20; f x = x - 9; main = f 10;",
            "offset": 33,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "x = 20; f x = x - 9; main = f 10;"
                    ]
                ]
            }
        }
    ]
}
